File: frontend/src/types/marketRegime.ts
// Market Regime Intelligence — frontend types (mirror backend schemas/market_regime.py)

export type MarketRegimeType =
  | 'TRENDING_BULL'
  | 'TRENDING_BEAR'
  | 'RANGE_BOUND'
  | 'HIGH_VOLATILITY'
  | 'LOW_VOLATILITY'
  | 'BREAKOUT'
  | 'REVERSAL'
  | 'UNKNOWN'

export type RegimeConfidence = 'LOW' | 'MEDIUM' | 'HIGH'
export type RegimeStatus = 'FAVORABLE' | 'NEUTRAL' | 'UNFAVORABLE'

export interface RegimePerformance {
  regime: MarketRegimeType
  sample_size: number
  wins: number
  losses: number
  breakeven: number
  win_rate: number | null
  avg_r: number | null
  expectancy_r: number | null
  profit_factor: number | null
  total_pnl: number | null
  confidence: RegimeConfidence
  status: RegimeStatus
}

export interface RegimePerformanceResponse {
  generated_at: string
  regimes: RegimePerformance[]
  matched_trades: number
  favorable_regimes: MarketRegimeType[]
  unfavorable_regimes: MarketRegimeType[]
}

export interface SetupRegimeCell {
  regime: MarketRegimeType
  sample_size: number
  avg_r: number | null
  expectancy_r: number | null
  win_rate: number | null
  confidence: RegimeConfidence
}

export interface SetupRegimeRow {
  setup: string
  cells: SetupRegimeCell[]
  best_regime: MarketRegimeType | null
  worst_regime: MarketRegimeType | null
}

export interface SetupRegimeMatrix {
  generated_at: string
  regimes: MarketRegimeType[]
  rows: SetupRegimeRow[]
}

export interface CurrentRegime {
  regime: MarketRegimeType
  confidence: RegimeConfidence
  as_of_date: string | null
  status: RegimeStatus
  reasoning: string[]
  nifty_trend: string | null
  nifty_regime: string | null
  nifty_change_pct: number | null
  india_vix: number | null
  atr_pct: number | null
  advance_decline_ratio: number | null
  best_setup: string | null
  best_setup_expectancy_r: number | null
  worst_setup: string | null
  worst_setup_expectancy_r: number | null
}

export interface MarketRegimeDashboard {
  generated_at: string
  current: CurrentRegime | null
  performance: RegimePerformanceResponse
  matrix: SetupRegimeMatrix
}

export const REGIME_LABELS: Record<MarketRegimeType, string> = {
  TRENDING_BULL: 'Trending Bull',
  TRENDING_BEAR: 'Trending Bear',
  RANGE_BOUND: 'Range Bound',
  HIGH_VOLATILITY: 'High Volatility',
  LOW_VOLATILITY: 'Low Volatility',
  BREAKOUT: 'Breakout',
  REVERSAL: 'Reversal',
  UNKNOWN: 'Unknown',
}

File: backend/app/schemas/market_regime.py
"""Pydantic schemas for the Market Regime Intelligence engine.

All metrics are derived from existing MarketSnapshot fields + closed trade
R-multiples. No new indicators, no LLM scoring. User-scoped, read-only.
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class MarketRegimeType(str, Enum):
    TRENDING_BULL = "TRENDING_BULL"
    TRENDING_BEAR = "TRENDING_BEAR"
    RANGE_BOUND = "RANGE_BOUND"
    HIGH_VOLATILITY = "HIGH_VOLATILITY"
    LOW_VOLATILITY = "LOW_VOLATILITY"
    BREAKOUT = "BREAKOUT"
    REVERSAL = "REVERSAL"
    UNKNOWN = "UNKNOWN"


class RegimeConfidence(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class RegimeStatus(str, Enum):
    FAVORABLE = "FAVORABLE"
    NEUTRAL = "NEUTRAL"
    UNFAVORABLE = "UNFAVORABLE"


class RegimePerformance(BaseModel):
    regime: MarketRegimeType
    sample_size: int = 0
    wins: int = 0
    losses: int = 0
    breakeven: int = 0
    win_rate: Optional[float] = None
    avg_r: Optional[float] = None
    expectancy_r: Optional[float] = None
    profit_factor: Optional[float] = None
    total_pnl: Optional[float] = None
    confidence: RegimeConfidence = RegimeConfidence.LOW
    status: RegimeStatus = RegimeStatus.NEUTRAL


class RegimePerformanceResponse(BaseModel):
    generated_at: str
    regimes: list[RegimePerformance] = Field(default_factory=list)
    matched_trades: int = 0
    favorable_regimes: list[MarketRegimeType] = Field(default_factory=list)
    unfavorable_regimes: list[MarketRegimeType] = Field(default_factory=list)


class SetupRegimeCell(BaseModel):
    regime: MarketRegimeType
    sample_size: int = 0
    avg_r: Optional[float] = None
    expectancy_r: Optional[float] = None
    win_rate: Optional[float] = None
    confidence: RegimeConfidence = RegimeConfidence.LOW


class SetupRegimeRow(BaseModel):
    setup: str
    cells: list[SetupRegimeCell] = Field(default_factory=list)
    best_regime: Optional[MarketRegimeType] = None
    worst_regime: Optional[MarketRegimeType] = None


class SetupRegimeMatrix(BaseModel):
    generated_at: str
    regimes: list[MarketRegimeType] = Field(default_factory=list)
    rows: list[SetupRegimeRow] = Field(default_factory=list)


class CurrentRegime(BaseModel):
    regime: MarketRegimeType
    confidence: RegimeConfidence = RegimeConfidence.LOW
    as_of_date: Optional[str] = None
    status: RegimeStatus = RegimeStatus.NEUTRAL
    reasoning: list[str] = Field(default_factory=list)
    # Raw signals used for classification (existing snapshot fields)
    nifty_trend: Optional[str] = None
    nifty_regime: Optional[str] = None
    nifty_change_pct: Optional[float] = None
    india_vix: Optional[float] = None
    atr_pct: Optional[float] = None
    advance_decline_ratio: Optional[float] = None
    # Best/worst setup for the active regime (data-driven)
    best_setup: Optional[str] = None
    best_setup_expectancy_r: Optional[float] = None
    worst_setup: Optional[str] = None
    worst_setup_expectancy_r: Optional[float] = None


class MarketRegimeDashboard(BaseModel):
    generated_at: str
    current: Optional[CurrentRegime] = None
    performance: RegimePerformanceResponse
    matrix: SetupRegimeMatrix
